rustler_tests/native/dynamic_load/src/lib.rs

```diff
@@ -0,0 +1,40 @@
+use rustler::Atom;
+use std::{ffi::OsStr, fs::read_to_string, os::unix::prelude::OsStrExt, path::PathBuf};
+
+static mut DATASET: Option<Box<str>> = None;
+
+fn initialize_dataset(mut asset_path: PathBuf) {
+    asset_path.push("demo_dataset.txt");
+
+    // https://github.com/elixir-lsp/elixir-ls/issues/604
+    eprintln!("Loading dataset from {:?}.", &asset_path);
+
+    let data = read_to_string(asset_path).unwrap().into_boxed_str();
+    let data = Some(data);
+
+    // Safety: assumes that this function is being called once when
+    // dynamically loading this library.
+    // `load()` is being called exactly once and OTP will not allow any other function call
+    // before this function returns
+    unsafe { DATASET = data };
```

Comment:
I think we can remove the need for unsafe here if we use `once_cell`?

Reply:
Relying on external synchronization mechanisms is unsafe by definition. We need to use unsafe here because rust compiler doesn't know about synchronization guarantees provided by OTP's NIF implementation.

If we were to use any safe rust code it would need to provide its own synchronization which defeats the purpose of relying on OTP's synchronisation or lie to the compiler about synchronization which defeats the purpose of unsafe {} keyword.

Also [std::cell::OnceCell](https://doc.rust-lang.org/std/cell/struct.OnceCell.html) is in nightly so it's a no go anyway.

Reply:
> If we were to use any safe rust code it would need to provide its own synchronization which defeats the purpose of relying on OTP's synchronisation

But that would be ok, wouldn't it? As long as we can guarantee that the value is only written once, we are safe I think?

I wonder if we can somehow forbid that `DATASET` is written twice, without using `once_cell`. But we can probably revisit that later? Note that examples from the tests might be reused by users.

Reply:
@scrogson I am ok with using unsafe here. `once_cell` being nightly is enough reason for me to stick with `unsafe`. What do you think?

Reply:
> I wonder if we can somehow forbid that DATASET is written twice, without using once_cell. But we can probably revisit that later? Note that examples from the tests might be reused by users.

We could put `DATASET` as private member of some module and expose `public(crate)` api for initialization and reading.

Regarding synchronization done inside the rust code we could in theory use [std::sync::RwLock](https://doc.rust-lang.org/stable/std/sync/struct.RwLock.html) whoever it can be poisoned and [LockResult](https://doc.rust-lang.org/stable/std/sync/type.LockResult.html)<[RwLockReadGuard](https://doc.rust-lang.org/stable/std/sync/struct.RwLockReadGuard.html)<'_, T>> also doesn't look encouraging.

If we were to ensure the DATASET is only initialized once we could use [std::sync::atomic:AtomicPtr::compare_exchange_weak](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicPtr.html#method.compare_exchange_weak) with `current` param as `null_ptr` however `pointer::as_ref` is also `unsafe`. It gives us a runtime guarantee however we still need to use unsafe keyword.

Reply:
The atomicity of `AtomicPtr` seems preferable, even though `on_load` should be called only once anyways. I think with `AtomicPtr`, we also don't need to hide the data, writing to the `AtomicPtr` does not happen by accident. @kaaboaye what do you think, should we use `AtomicPtr` or is the current version prettier?

Reply:
Frankly I would stick with Option. When upgrade happens it will cause memory leak instead of crash which I think will be preferable during development. Assuming there are no upgrades done in production environment.

Reply:
As long as load is even being called multiple times. I have no idea